website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,651 @@
+
+When comparing BSON values of different types, the BSON comparison order is used,
+with array being an exception.
+When comparing values of the same type, the specific comparison for that type is used.
+
+### Comparing Arrays
+
+When comparing BSON arrays, each index is iterated in order.
+At each index, BSON comparison order is used when the values are compared.
+If the values at a particular index are not equal, that result is used as the final comparison result.
+
+If there are no more elements in the array during the iteration,
+then that array is considered smaller.
+
+#### Example of Comparing Arrays
+
+To compare two arrays, we compare each element in the arrays until we discover a difference.
+For instance, when comparing the arrays `[3, 1, 'b']` and `[3, 'b', 'a']`,
+the following steps are taken:
+
+1. Compare `[3, 1, 'b']` and `[3, 'b', 'a']` using comparison order.
+   They have the same BSON type; both are arrays.
+2. Compare first index `3` and `3` using comparison order, both are numbers, and they are the same.
+3. Compare next index, `1` and `'b'` using comparison order,
+`1` is a number and less than string `'b'`.
+
+So `[3, 1, 'b']` is less than `[3, 'b', 'a']`.
+
+If an array has no more element to compare, the one with no more element is less.
+For the same reason empty array is less than arrays with elements.
+
+### Comparing Array and Other BSON Type
+
+When comparing an array with other BSON types, it doesn't *always* use BSON comparison order
+of array BSON type.
+It often uses the greatest or the smallest value
+from the array to compare against other BSON type.
+
+In some cases, such as with the [`$max` and `$min` update operator](#comparison-in-update-operator),
+the comparison order is used.
+In other situations, such as [sorting](#comparison-in-sorting),
+the [`$gt` and `$lt` filter operators](#comparison-in-filter-operator),
+an element within the array is used to compare against other BSON type.
+
+[Sorting](#comparison-in-sorting), used in `.sort()`, uses the following rules
+to compare array and other BSON types.
+
+* Ascending sort uses the smallest element of any BSON type from the array.
+* Descending sort uses the largest element of any BSON type from the array.
+* Empty array is less than null.
+
+[Filtering](#comparison-in-filter-operator), commonly used in `$lt` and `$gt` operators,
+applies the following rules to compare array and other BSON types.
+
+* `$lt` uses the smallest element of the same BSON type as non-array from the array.
+* `$gt` uses the largest element of the same BSON type as non-array from the array.
+* Empty array should not match other BSON type.
+
+[Updating](#comparison-in-update-operator) with `$max` and `$min` operators applies the following rules
+to compare array and other BSON types.
+
+* For array and other BSON type comparison, BSON comparison order is used.
+* Empty array is greater than null using BSON comparison order.
+
+#### Example of Ascending and Descending Sort
```

Comment:
It looks a bit strange that we give 3 sets of rules, but then give an example only for the first one.

I think it might be better to add short examples for all sets:

* sorting rules with small examples
* filtering rules with small examples
* updating rules with small examples
* demonstration with mongosh